src/main/java/org/acme/consumer/OSSIndexBatcher.java

```diff
@@ -0,0 +1,101 @@
+package org.acme.consumer;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+import io.quarkus.kafka.client.serialization.ObjectMapperSerde;
+
+import io.quarkus.runtime.StartupEvent;
+import org.acme.event.OssIndexAnalysisEvent;
+import org.acme.event.VexUploadEvent;
+import org.acme.model.Component;
+import org.acme.tasks.scanners.OssIndexAnalysisTask;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.*;
+
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Properties;
+import java.util.concurrent.LinkedBlockingQueue;
+
+@ApplicationScoped
+public class OSSIndexBatcher {
```

Comment:
Thanks for the input Alioune..this is done

Reply:
Hi Alioune....for OSS Index Batcher, we are actually calling the processor after the capacity is reached to achieve the pagination we were doing in DT and punctuator is implemented in case we dont received the mentioned capacity by the time period then we should go ahead and call the OSS API with available number of components.
As I am going through windowed aggregation that you mentioned, I see grouping only on time window, can we also do count based grouping using windowed aggregation?

Also, we are not planning to do aggregation for Snyk analyzer, that is why the batching logic is implemented in OSSIndexBatcher only and not in PrimaryConsumer.

Would look for your further input on this.

Reply:
> Hi Alioune....for OSS Index Batcher, we are actually calling the processor after the capacity is reached to achieve the pagination we were doing in DT and punctuator is implemented in case we dont received the mentioned capacity by the time period then we should go ahead and call the OSS API with available number of components. As I am going through windowed aggregation that you mentioned, I see grouping only on time window, can we also do count based grouping using windowed aggregation?

Grouping can only be done by time window. I Think that only timed aggregation could be enough. Both approaches cover the batching requirements, so way have a look at the drawback 

1. With capacity-based grouping, the processor have to know two informations : the max capacity and the time window. You need to make sure you don't get stuck into your processor if there are not enough components hence your punctuator. Because the grouping is done beforehand, you removed the pagination from `OssIndexAnalysisTask` but I think it is mostly the responsibility of each analyzer to know about its pagination method. `OssIndexAnalysisTask` should continue to paginate because you cannot enforce that Kafka Streams will always be its source and the data will be correctly paginated before reaching the analyzer ==> Responsibilities are mixed here or a tight coupling is supposed. Moreover, you cannot handle the situation where you have batched together 100 components and 98 of them are still in the cache ==> In the end, you are going to send a batch of 2 components. 
2. With time-based grouping, Kafka Streams only have to know about the time windows. You need however to handle the case where you aggregated much more components than you wanted but that's not a problem if the analyzer is self-sufficient in its batching (that's was already the case). Responsibilities are clear and you can handle the cache case, if you receive a group of 250 components with 150 still in the cache, you can filter out those still in cache before building on full batch instead of 3 incomplete batches (better rate limiting handling). You can also call the analyzer task outside the context of Kafka Streams topology and still have it perform the pagination

Also even though, you are not currently doing batch call for Snyk because the API does not expose such feature, all analyzers are designed to take a list of component (i.e `void analyze(List<Component> components)`). The component will be more flexible if you handle snyk specifics in the task rather than upstream in the Kafka Streams topology.
Easier to handle a pagination of 1 than to implement pagination afterward.

Reply:
This is implemented

Reply:
Thanks for the inputs and explanation..This has been implemented